contracts/contracts/coordination/GlobalAllowList.sol

```diff
@@ -15,63 +20,143 @@ contract GlobalAllowList is IEncryptionAuthorizer {
+     * @return The authorization status
+     */
     function isAuthorized(
         uint32 ritualId,
         bytes memory evidence,
         bytes memory ciphertextHeader
     ) external view override returns (bool) {
+        _beforeIsAuthorized(ritualId, evidence, ciphertextHeader);
+
         bytes32 digest = keccak256(ciphertextHeader);
         address recoveredAddress = digest.toEthSignedMessageHash().recover(evidence);
         return isAddressAuthorized(ritualId, recoveredAddress);
     }
 
+    /**
+     * @dev This function is called before the setAuthorizations function
+     * @param ritualId The ID of the ritual
+     * @param addresses The addresses to be authorized
+     * @param value The authorization status
+     */
+    function _beforeSetAuthorization(
+        uint32 ritualId,
+        address[] calldata addresses,
+        bool value
+    ) internal view virtual {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    /**
+     * @notice Authorizes a list of addresses for a ritual
+     * @param ritualId The ID of the ritual
+     * @param addresses The addresses to be authorized
+     */
     function authorize(
         uint32 ritualId,
         address[] calldata addresses
-    ) external onlyAuthority(ritualId) {
+    ) external canSetAuthorizations(ritualId) {
         setAuthorizations(ritualId, addresses, true);
     }
 
+    /**
+     * @notice Deauthorizes a list of addresses for a ritual
+     * @param ritualId The ID of the ritual
+     * @param addresses The addresses to be deauthorized
+     */
     function deauthorize(
         uint32 ritualId,
         address[] calldata addresses
-    ) external onlyAuthority(ritualId) {
+    ) external canSetAuthorizations(ritualId) {
         setAuthorizations(ritualId, addresses, false);
     }
 
+    /**
+     * @notice Sets the authorization status for a list of addresses for a ritual
+     * @dev Only active rituals can set authorizations
+     * @param ritualId The ID of the ritual
+     * @param addresses The addresses to be authorized or deauthorized
+     * @param value The authorization status
+     */
     function setAuthorizations(uint32 ritualId, address[] calldata addresses, bool value) internal {
-        require(coordinator.isRitualActive(ritualId), "Only active rituals can add authorizations");
+        require(coordinator.isRitualActive(ritualId), "Only active rituals can set authorizations");
+
+        require(addresses.length <= MAX_AUTH_ACTIONS, "Too many addresses");
+
+        _beforeSetAuthorization(ritualId, addresses, value);
         for (uint256 i = 0; i < addresses.length; i++) {
-            authorizations[lookupKey(ritualId, addresses[i])] = value;
+            authorizations[LookupKey.lookupKey(ritualId, addresses[i])] = value;
             emit AddressAuthorizationSet(ritualId, addresses[i], value);
         }
+
+        authActions[ritualId] += addresses.length;
```

Comment:
Is incrementing only applicable if `value` is `True`? If `value` is `False` should this be decrementing instead?

Reply:
For this version, the goal was to charge per action, regardless if it's authorizing or deauthorizing. But more recent variations proposed to some adopters will use something different (e.g. charging for authorized slots). I think we can leave it as is because it's only a starting point anyway, and the more recent design of subscription contracts will refactor everything

Reply:
👍 I see, so every action (authorize or deauthorize) is a charge.